src/main/java/me/oyurimatheus/nossomercadolivre/products/Product.java

```diff
@@ -194,14 +195,17 @@ private void greaterThanZero(BigDecimal price, String msg) {
     /**
      * changes the product {@link #stockQuantity}
      *
-     * @param purchase a new Purchase
+     * @param newPurchase a new Purchase
+     * @param buyer a buyer
      */
-    public void reserveQuantityFor(Purchase purchase) {
-        if (stockQuantity < purchase.getQuantity()) {
-            throw new IllegalStateException("Quantity required is unavailable");
+    public Optional<Purchase> reserveQuantityFor(NewPurchaseRequest newPurchase, User buyer) {
```

Comment:
*Operações sobre dados cujo tipo é padrão da linguagem devem ficar dentro das nossas classes.*

Uma dúvida... Pq você não recebe só um int aqui com a quantidade a ser abatida? Você opera sobre a quantidade da ```newPurchase``` que é um int... Só para deixar claro, acho que estaria tudo bem. Mas isso só aconteceu pq você passou um tipo complexo aqui quando poderia ter passado um int :). 

Meu código teria ido na linha de fazer com que o reserveQuantity me retornasse um objeto do tipo NewReserveQuantity(quantidade,produto) e isso seria usado para construir uma nova purchase. Aqui você criaria uma classe a mais, mas com o objetivo de usar como um objeto de transição. Pena que não temos tuplas ou outra  no java :).

Reply:
Porque eu queria forçar que a reserva fosse feita em cima da quantidade da `newPurchase`... Se eu recebesse um `int`, pode ser que o valor passado pelo parâmetro, não seja o mesmo da requisição, aí teria que validar isso na hora da criação da compra.

Reply:
Daqui de dentro desse método você não tem essa visão né? Como que você sabe que o valor passado é a newPurchase que chegou pelo controller? O método só enxerga seu escopo, nada fora. Então minha sugestão é focar no que tem controle. Você pode pedir um int como argumento e retornar a combinação. 

E claro, se achar melhor seguir desse jeito, jogue duro.

Reply:
Justo justo, vou ver no próximo PR, qualquer coisa refatoro